operator/rediscontroller/observe.go

```diff
@@ -0,0 +1,169 @@
+	}
+	redis := *resp.JSON200
+
+	log.V(2).Info("response", "raw", string(resp.Body))
+	log.V(1).Info("retrieved instance", "state", redis.State)
+
+	redisInstance.Status.AtProvider, err = mapObservation(redis)
+	if err != nil {
+		log.Error(err, "unable to fully map observation, ignoring.")
+	}
+
+	var state oapi.EnumServiceState
+	if redis.State != nil {
+		state = *redis.State
+	}
+	switch state {
+	case oapi.EnumServiceStateRunning:
+		redisInstance.SetConditions(exoscalev1.Running())
+	case oapi.EnumServiceStateRebuilding:
+		redisInstance.SetConditions(exoscalev1.Rebuilding())
+	case oapi.EnumServiceStatePoweroff:
+		redisInstance.SetConditions(exoscalev1.PoweredOff())
+	case oapi.EnumServiceStateRebalancing:
+		redisInstance.SetConditions(exoscalev1.Rebalancing())
+	default:
+		log.V(2).Info("ignoring unknown instance state", "state", state)
+	}
+
+	res, err := p.exo.GetDbaasCaCertificateWithResponse(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, fmt.Errorf("unable to retrieve CA certificate: %w", err)
+	}
+	ca := *res.JSON200.Certificate
+
+	rp, err := mapParameters(redis, redisInstance.Spec.ForProvider.Zone)
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+	cd, err := connectionDetails(redis, ca)
+	if err != nil {
+		return managed.ExternalObservation{}, fmt.Errorf("unable to parse connection details: %w", err)
+	}
+
+	observation := managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        isUpToDate(&redisInstance.Spec.ForProvider, rp, log),
+		ResourceLateInitialized: false,
+		ConnectionDetails:       cd,
+	}
+
+	return observation, nil
+}
+
+func isUpToDate(current, external *exoscalev1.RedisParameters, log logr.Logger) bool {
+	if external == nil {
+		return false
+	}
+	extIPFilter := []string(external.IPFilter)
+	checks := map[string]bool{
+		"IPFilter":              mapper.IsSameStringSet(current.IPFilter, &extIPFilter),
+		"Maintenance":           current.Maintenance.Equals(external.Maintenance),
+		"Size":                  current.Size.Equals(external.Size),
+		"TerminationProtection": current.TerminationProtection == external.TerminationProtection,
+		"RedisSettings":         mapper.CompareSettings(current.RedisSettings, external.RedisSettings),
+	}
+	ok := true
+	for _, v := range checks {
+		if !v {
+			log.V(2).Info("instance not up-to-date", "check", v)
+			ok = false
+		}
+	}
+	return ok
+}
+
+func connectionDetails(in oapi.DbaasServiceRedis, ca string) (map[string][]byte, error) {
+	if in.ConnectionInfo == nil || len(*in.ConnectionInfo.Uri) == 0 {
+		return map[string][]byte{}, nil
+	}
+	uri := (*in.ConnectionInfo.Uri)[0]
+	parsed, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	password, _ := parsed.User.Password()
+	return map[string][]byte{
+		"host":      []byte(parsed.Hostname()),
+		"port":      []byte(parsed.Port()),
+		"username":  []byte(parsed.User.Username()),
+		"password":  []byte(password),
```

Comment:
Any specific reasons why these are lowercased? They reek of upper-cased env variables :) Aren't there some commonly named variables in redis-cli or so?

Reply:
that's based on what's written in the ticket. I explicitly added REDIS_URL because that's what is usually necessary. The rest is there because of the ticket description.